Design/Pre-commit/Pre-commit implementation detail.md

```diff
@@ -0,0 +1,110 @@
+ 
+#### XML and YAML checking 
+- Should catch yaml and xml violations in commits before they become a problem, on the build servers. 
+ 
+#### Cppcheck 
+- Run Cppcheck with errors displayed on the command line and output 
+    - Cppcheck will only be present if the suppression of warnings is possible to implement simply, if not it will need to be added at a later date. 
+    - We will be maintaining this hook as we can then control the cppcheck version 
+ 
+#### Pre-commit run order 
+- Check XML/YAML 
+- Yapf - In place fixes 
+- Clang-format - In place fixes 
+- Flake8 
+- Cppcheck (Provisionally) 
+ 
+An initial implementation has been made here: https://github.com/mantidproject/mantid/pull/30740 
+ 
+## Installation instructions 
+ 
+We will be installing pre-commit and setting it up with the git hooks so that it runs before each time you commit. 
+ 
+#### CentOS/RHEL Installation 
+ 
+To install snap for CentOS and RHEL to support the use of pre-commit, without using pip. Navigate to your repo and run: 
+```shell 
+pip install pre-commit 
+pre-commit install 
+``` 
+ 
+#### Ubuntu 18.04+ Installation 
+Navigate to your Repo and run: 
+```shell 
+sudo snap install pre-commit --classic 
+pre-commit install 
+``` 
+ 
+### MacOS Installation 
+We install pre-commit via homebrew on MacOS. Navigate to your repo and run: 
+```shell 
+sudo brew install pre-commit 
+pre-commit install 
+``` 
+ 
+### Windows Installation 
+On Windows this is installed and initialised for them via CMake, just pull the latest changes from master, re-run CMake and it will sort it for you. 
+ 
+## Migration steps 
+- Create a separate repo called mantidproject/pre-commit-hooks e.g. https://github.com/pocc/pre-commit-hooks 
+    - Add clang-format 
+    - Add cppcheck (provisionally, if too complex will leave to a later date) 
+- Merge pre-commit framework PR (https://github.com/mantidproject/mantid/pull/30740) 
+- Create pre-commit CI Job that will replace: 
+    - Clang-format 
+    - flake8 
+    - cppcheck (If implemented at earlier step) 
+ 
+### Final Migration Steps (TBD) 
+All final tasks should be completed in one day on to minimise potential disruption to developers in the team 
+- Update and apply clang-format and Yapf versions to the whole codebase, using the pre-commit hooks 
+- Replace Jenkins static analysis jobs with the already made pre-commit CI job 
+- Inform developers
+
+## Potential future feature plans
+Future development will be easy and not require significant updates to CMake for Windows or require custom installation, updating the .yaml config will allow future changes to not only the local developer machines but on Jenkins seamlessly.
+
+- Any newly added feature will be run on the whole repo at the same time to avoid random things being added to new PRs and ideally should be enforced by Jenkins, but many suggested features won't require jenkins enforcement. 
+- Support more things in pre-commit slowly introducing helpful features 
```

Comment:
Add key and password finding to this list also, on the off chance someone thinks that is a good idea.

Reply:
Thought I'd added this, I am keenly aware of the issues with key and password leakage having done that myself on a personal project (Never again!)